### PR TITLE
use explicit stack to prevent stack overflow in drop specialization

### PR DESCRIPTION
**Pull request by folkertdev**

### CODE REVIEW

File: crates/compiler/mono/src/drop_specialization.rs [hunk 1]
```diff
@@ -86,167 +86,198 @@ fn specialize_drops_stmt<'a, 'i>(
...
-                            ..
-                        } => {
-                            let [structure, index] = match arguments {
-                                [structure, index] => [structure, index],
-                                _ => unreachable!("List get should have two arguments"),
-                            };
+            let Call {
+                call_type,
+                arguments,
+            } = call;
```
[the remark this answers is not in the document]
Comment (folkertdev) [reply crates/compiler/mono/src/drop_specialization.rs]: done

File: crates/compiler/mono/src/drop_specialization.rs [hunk 2]
```diff
@@ -86,167 +86,193 @@ fn specialize_drops_stmt<'a, 'i>(
...
+                            Struct(children) => {
+                                for (index, child) in children.iter().enumerate() {
+                                    environment.add_struct_child(*binding, *child, index as u64);
+                                }
+                            }
+                            StructAtIndex {
+                                index, structure, ..
+                            } => {
+                                environment.add_struct_child(*structure, *binding, *index);
+                                // alloc_let_with_continuation!(environment)
```
Comment (JTeeuwissen) [inline crates/compiler/mono/src/drop_specialization.rs:198]: this can be removed?

File: crates/compiler/mono/src/drop_specialization.rs [hunk 3]
```diff
@@ -86,167 +86,193 @@ fn specialize_drops_stmt<'a, 'i>(
...
+            use Expr::*;
+
+            // to prevent stack overflows, try to use an explicit stack to accumulate a bunch of
+            // Let statements. Call expressions require more logic and are never put on this stack
+            let mut stack = vec![];
+
+            let mut stmt = stmt;
+
+            let final_continuation = loop {
+                match stmt {
```
Comment (JTeeuwissen) [inline crates/compiler/mono/src/drop_specialization.rs:156]: I'd prefer a 
```rs
while let Stmt::Let(binding, expr, layout, next_stmt) = stmt
```

File: crates/compiler/mono/src/drop_specialization.rs [hunk 4]
```diff
@@ -86,167 +86,193 @@ fn specialize_drops_stmt<'a, 'i>(
...
-                    // Does nothing relevant to drop specialization. So we can just continue.
-                    alloc_let_with_continuation!(environment)
-                }
+            for (binding, expr, layout) in stack.into_iter().rev() {
+                final_continuation = arena.alloc(Stmt::Let(
+                    binding,
+                    expr,
+                    layout,
+                    arena.alloc(final_continuation),
+                ));
```
Comment (JTeeuwissen) [inline crates/compiler/mono/src/drop_specialization.rs:272]: I'd prefer a fold
```rs
let final_continuation = stack.into_iter().rev().fold(
    specialize_drops_stmt(
        arena,
        layout_interner,
        ident_ids,
        environment,
        final_continuation,
    ),
    |final_continuation, (binding, expr, layout)| {
        arena.alloc(Stmt::Let(
            binding,
            expr,
            layout,
            arena.alloc(final_continuation),
        ))
    },
);
```

Comment (JTeeuwissen) [reply crates/compiler/mono/src/drop_specialization.rs:272]: And perhaps update the name as well. If it's not really final.

**Review by JTeeuwissen — CHANGES_REQUESTED** [review]

**Review by JTeeuwissen — APPROVED** [review]

**Pull request merged**